// redux-form/es/structure/immutable/index.js
import _toPath from 'lodash-es/toPath';
import { Map, Iterable, List, fromJS as _fromJS } from 'immutable';

import deepEqual from './deepEqual';
import plainGetIn from '../plain/getIn';

var structure = {
  empty: Map(),
  getIn: function getIn(state, field) {
    return Map.isMap(state) || List.isList(state) ? state.getIn(_toPath(field)) : plainGetIn(state, field);
  },
  setIn: function setIn(state, field, value) {
    return state.setIn(_toPath(field), value);
  },
  deepEqual: deepEqual,
  deleteIn: function deleteIn(state, field) {
    return state.deleteIn(_toPath(field));
  },
  fromJS: function fromJS(jsValue) {
    return _fromJS(jsValue, function (key, value) {
      return Iterable.isIndexed(value) ? value.toList() : value.toMap();
    });
  },
  size: function size(list) {
    return list ? list.size : 0;
  },
  some: function some(iterable, callback) {
    return Iterable.isIterable(iterable) ? iterable.some(callback) : false;
  },
  splice: function splice() {
    var list = arguments.length <= 0 || arguments[0] === undefined ? List.isList(list) || List() : arguments[0];
    var index = arguments[1];
    var removeNum = arguments[2];
    var value = arguments[3];

    if (removeNum) {
      return list.splice(index, removeNum); // removing
    }
    if (index < list.size) {
      return list.splice(index, 0, value); // inserting
    }
    return list.set(index, value); // inserting after end of list
  }
};

export default structure;
